src/sage/categories/finite_fields.py

```diff
@@ -243,5 +246,232 @@ def _element_of_factored_order(self, F):
+                Traceback (most recent call last):
+                ...
+                ValueError: there are no non-squares in finite fields of even order
+            """
+            # if the order is an even power of two
+            # then every element is a square
+            if self.characteristic() == 2:
+                raise ValueError("there are no non-squares in finite fields of even order")
+            for element in self:
+                if not element.is_square():
+                    return element
+
     class ElementMethods:
-        pass
+        def is_square(self) -> bool:
+            r"""
+            Tests if the element is a square or has
+            a square root element.
+
+            OUTPUT: ``True`` if the element is a square ``False`` if not
+
+            EXAMPLES::
+
+                sage: S.<x> = GF(5)[]
+                sage: f = S.irreducible_element(20)
+                sage: k.<y> = S.quotient_ring(f)
+                sage: k in Fields()
+                True
+                sage: k(2).is_square()
+                True
+                sage: k.quadratic_non_residue().is_square()
+                False
+            """
+            if self.is_zero():
+                return True
+            if self.parent().characteristic() == 2:
+                return True
+            q = self.parent().order()
+            character = self**((q-1)//2)
+            is_square = character == self.parent().one()
+            return is_square
+
+        def _tonelli(self):
+            r"""
+            Returns a square root of the element if it exists
+            using Tonelli's algorithm, only works for finite fields
+            of odd characteristic.
+
+            OUTPUT: A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
+
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._tonelli()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._tonelli()
+                Traceback (most recent call last):
+                ...
+                ValueError: element is not a square
+            """
+            q = self.parent().cardinality()
+            if not self.is_square():
+                raise ValueError("element is not a square")
+            g = self.parent().quadratic_non_residue()
+            even_exp, odd_order = (q - ZZ.one()).val_unit(2)
+            e = 0
+            for i in range(2, even_exp+1):
+                tmp = self * (pow(g, -e))
+
+                condition = tmp**((q-1)//(2**i)) != self.parent().one()
+                if condition:
+                    e = 2**(i-1) + e
+            h = self * (g**(-e))
+            b = g**(e//2) * h**((odd_order+1)//2)
+            return b
+
+        def _cipolla(self):
+            r"""
+            Returns a square root of the element if it exists
```

Comment:
First line should use imperative mode, so "Return" and not "Returns". Same elsewhere in the new methods.